[DOCS-14323/DOCS-14514/DOCS-14324] Add metrics GA processors (#37656)
DataDog/documentation: content/en/observability_pipelines/processors/tag_cardinality_control.md

---
title: Tag Cardinality Control Processor
disable_toc: false
products:
- name: Metrics
  icon: metrics
  url: /observability_pipelines/configuration/?tab=metrics#pipeline-types
---

{{< product-availability >}}

{{< callout url="#"
 btn_hidden="true" header="Join the Preview!">}}
The Tag Cardinality Control processor is in Preview. Contact your account manager to request access.
{{< /callout >}}

## Overview

The Tag Cardinality Control processor limits the number of tag values for each metric. For example, a metric with unbounded tag keys, such as `userID`, can cause the metric's cardinality to spike and impact ingestion and indexing costs. To prevent these unexpected spikes, use the processor to set a cardinality limit for metrics that match the filter query, and either drop metrics received after the limit is reached or drop the tags for those metrics.

> Optionally, you can also configure [per-metric overrides](#optional-per-metric-override-settings) to set a limit for a specific metric or to exclude the metric from any cardinality limits. For each per-metric override, you can also set a custom limit for individual tags within the metric, or exclude the tag from the per-metric cardinality limit.

Reviewer: ```suggestion
Optionally, you can also configure [per-metric overrides](#optional-per-metric-override-settings) to set a limit for a specific metric or to exclude the metric from any cardinality limits. For each per-metric override, you can also set a custom limit for individual tags within the metric, or exclude them from the per-metric cardinality limit.
```

## Setup

To set up the Tag Cardinality Control processor:

> 1. Define a filter query. Only matching metrics are processed by this processor. See [Metrics Search Syntax][1] for more information.

Reviewer: ```suggestion
1. Define a filter query. See [Metrics Search Syntax][1] for more information.
    - Only metrics matching the filter are processed.. 
    - All metrics, regardless of whether they match the filter query, are sent to the next step in the pipeline.
```

Reviewer: Likewise suggesting here that the detail about only matching metrics being processed might be more effectively surfaced as a sub-bullet

1. In the **When the limit is reached** dropdown menu, select whether to **Drop tag** or **Drop event** for metrics that have exceeded the cardinality limit.

{{< img src="observability_pipelines/processors/tag_cardinality_control_settings.png" alt="The Tag Cardinality Control processor set to a cardinality limit of 200 and to drop tag when the limit is reached." style="width:40%;" >}}

### Optional per-metric override settings

If you want to set a cardinality limit for a specific metric:

1. Click **Manage Overrides** and then **Add Metric Override** in the side panel.
1. Enter the name of the metric.
1. Select the override mode in the dropdown menu.
    - **Custom limit**: Sets a cardinality limit for this metric.
    - **Exclude from limit**: Excludes this metric from being counted toward the cardinality limit. This is useful when you are tracking an important metric and you don't want to drop any samples or tags due to a cardinality limit.
1. In the **When the limit is reached** dropdown menu, select whether to **Drop tag** or **Drop event** for metrics that have exceeded the cardinality limit.

> 1. If you want to add specific tag overrides for this metric:

Reviewer: I'm finding the multiple levels of nesting for this numbered step to be a bit difficult to scan (the **Note** is currently four levels deep). I tried a couple of alternative presentations in a local build, and attached screenshots of each below:

**First potential option:** _Convert the step with multiple levels of nesting into its own subsection:_

<img width="762" height="425" alt="alternative-subsection" src="https://github.com/user-attachments/assets/b4670131-12c9-4971-b45e-c73554d9698d" />

**Second potential option:** _Use the same subsection as above, but also add the **Custom limit** / **Exclude from limit** bullet pairs to small indented tables._ Looks like some other OP docs use indented tables under numbered steps as well ([example](https://docs-staging.datadoghq.com/may/op-metrics-ga-processors/observability_pipelines/configuration/install_the_worker/?interface=ui&platform=cloudformation&secrets_source=secrets_management&tab=apt#install-the-worker)).

<img width="682" height="905" alt="alternative-table" src="https://github.com/user-attachments/assets/7591f93f-50cf-4973-9a82-1b0842bce09b" />

Do you think either of these options might make the information a bit easier to parse? Open to your feedback — thanks in advance!

Author: Originally, I had the Per-Tag Overrides in a separate section...I don't remember why I decided to combine it. Let me try it out with it separated again. I'd rather just use the bullet points and not the table though.

Author: Updated in this commit: https://github.com/DataDog/documentation/pull/37656/changes/174442f8fb3228e0d3644a99f91136d4530c8de0

    1. Click **Add Tag Override**.

>     1. Enter the tag key on which to set a limit.

Reviewer: ```suggestion
    1. Enter the tag key you want to set a limit for.
```

Reviewer: Minor suggestion to be a bit less formal, and also for consistency with `edit_tags.md`, which uses similar constructions such as "Enter the name of the tag key you want to rename".

    1. Select the override mode in the dropdown menu:
        - **Custom limit**: Sets a limit on the number of unique values per tag. For example, if the tag limit is set to `5`, the first five tag values received are used.

>             - **Note**: The tag values are only reset with a Worker restart or a pipeline configuration update, even if the Tag Cardinality Control processor isn't updated.

Reviewer: ```suggestion
            - **Note**: Tag values persist until the Worker restarts or the pipeline configuration is updated. Any pipeline configuration update resets the tag values, even if the update doesn't modify the Tag Cardinality Control processor.
```

Reviewer: I found it a bit difficult to understand the contrast of the two ideas presented here. Suggesting this split for clarity, but I'm open to your feedback.

        - **Exclude from limit**: Excludes metrics with the specified tag from being counted toward the cardinality limit.

>     1. Enter the limit for the maximum number of tag keys.

Author: ```suggestion
    1. Enter the limit for the maximum number of tag key values.
```
